apollo-router-core/src/services/http_compat.rs

```diff
@@ -61,6 +65,52 @@ impl<T> DerefMut for Request<T> {
     }
 }
 
+impl<T: Hash> Hash for Request<T> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.inner.method().hash(state);
+        self.inner.version().hash(state);
+        self.inner.uri().hash(state);
+        // this assumes headers are in the same order
+        for (name, value) in self.inner.headers() {
+            name.hash(state);
+            value.hash(state);
+        }
```

Comment:
I really dislike HeaderMap. We can't even sort it... 

I suppose the downside of not spotting duplicates is that our query can't be de-duplicated. Not a bug, but defeats the purpose of de-duplicating.

We can do something a bit crafty here. `HeaderValue` does implement `Ord` and `HeaderName` always convert to `&str`, so:
```
        // Map Header names into &str so we can sort
        let mut tmp: Vec<(&str, &HeaderValue)> = self
            .inner
            .headers()
            .iter()
            .map(|(k, v)| (k.as_str(), v))
            .collect();
        tmp.sort();
        for (name, value) in tmp {
            name.hash(state);
            value.hash(state);
        }
```
would give us a consistent ordering for hashing purposes. I think we could do the same for Eq as well.

Reply:
we should refrain from sorting the headers, since their order can have an impact. Example: the `Accept` header can have multiple values, which can come either as comma separated in one header value, or as multiple separated `Accept` headers. In that second case, if we sort the headers, that might reorder the values and change the behaviour.

The assumption for the cache key here is that similar queries coming from the same client will have the same shape (same user agent, same  list of headers in the same order...).

What we could do though is decide on which headers we consider for the cache. Once we do that, we would get stronger guarantees, and could expose in the docs the issues around ordering. Could we explore that in another PR though?

Reply:
yes, that's definitely a good idea to revisit it and find a robust solution